Issue API warnings when workload names are not DNS labels

thockin opened the pull request:
Workload APIs like Deployment use their own name to ultimately generate pod names.  Pod names are ALLOWED to be DNS Subdomains, but when we use the pod name as the pod's hostname, we are sort of in conflict.  Setting the pod hostname (internally not thru API) to something like "foo.bar" is poorly defined and the `hostname` field doesn't allow it at all.

Fixes #104195
xref #106732

/kind feature
/kind api-change
/sig apps

#### Special notes for your reviewer:


Example:
```
$ k create -f /tmp/bad.yaml 
Warning: metadata.name: a DNS label is recommended: [a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]
```
I don't know if there is a convention for these error messages.


```release-note
Added warnings about workload resources (Pods, ReplicaSets, Deployments, Jobs, CronJobs, or ReplicationControllers) whose names are not valid DNS labels.
```

```docs
Creating Pods, ReplicaSets, Deployments, Jobs, CronJobs, or ReplicationControllers whose `metadata.name` field is *not* a valid [DNS label](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names) will now get [API warnings](https://blog.k8s.io/2020/09/03/warnings/).   All previously valid names and API operations are still allowed - these are only warnings.

For example, naming a Deployment `foo.bar` (a DNS subdomain) now returns a  warning, whereas `foobar` or `foo-bar` (DNS labels) do not.  For backwards compatibility, the Kubernetes API for these resources allows the `metadata.name` field to be a [DNS subdomain](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).  DNS labels are a subset of DNS subdomains.  All DNS label names are valid DNS subdomain names, but using a DNS subdomain that is not _also_ a valid DNS label (e.g. a name which includes dots) can produce unexpected behaviors when used as a pod's hostname.

See [Object Names and IDs](https://k8s.io/docs/concepts/overview/working-with-objects/names/) for more information.
```

thockin committed:
pod: API warn when name is not DNS label

thockin committed:
rc: API warn when name is not DNS label

thockin committed:
rs: API warn when name is not DNS label

thockin committed:
deployment: API warn when name is not DNS label

thockin committed:
job: API warn when name is not DNS label

thockin committed:
cronjob: API warn when name is not DNS label

aojea commented:
/assign @soltysh @liggitt

liggitt commented on pkg/registry/apps/deployment/strategy.go:
```diff
@@ -84,13 +86,18 @@ func (deploymentStrategy) PrepareForCreate(ctx context.Context, obj runtime.Obje
-	return pod.GetWarningsForPodTemplate(ctx, field.NewPath("spec", "template"), &newDeployment.Spec.Template, nil)
+	var warnings []string
+	if msgs := utilvalidation.IsDNS1123Label(newDeployment.Name); len(msgs) != 0 {
+		warnings = append(warnings, fmt.Sprintf("metadata.name: a DNS label is recommended: %v", msgs))
```
general question: would it be more useful to indicate _why_ a DNS label is recommended in the warning rather than the validation failure message? as a user, I'd wonder what is breaking or not working well if I ignore this warning.

thockin replied:
The only really concrete WHY is hostnames.  Should we say that in the warning?  If I finish https://github.com/kubernetes/kubernetes/pull/106732 it will matter more

liggitt replied:
I would find a message saying "pod hostnames may not work properly" helpful in deciding how much I cared about urgently fixing the warning on an existing thing

liggitt replied:
I guess this is only a warning on create, which we'd hope they would encounter while developing their workload manifest, and we wouldn't spam people on updates to existing things... that's good

thockin replied:
How about: 

```
$ k create -f /tmp/bad.yaml 
Warning: metadata.name: this will become part of Pods' names and default hostnames, which can result in surprising behavior: a DNS label is recommended: [a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]
```

liggitt replied:
that's better, but I think we can make it even clearer. If the primary difference between DNS label and DNS subdomain validation is that it can contain a `.`, can we simplify to:

```
if strings.Contains(newDeployment.Name, ".") {
   ...
   > `Warning: metadata.name: names containing '.' are not recommended; this name is used in pod names and default hostnames, and names containing '.' can result in surprising behavior`
}
```

thockin replied:
There are 2 primary criteria that are different: 1) dots; 2) max length.  We can check for (and warn on) those individually.  In fact, looking at `IsDNS1123Label()` it does check length independent of regex.  We could make that do something like:

```
errs := []
if too long {
  errs += "too long"
}
if valid subdomain but not valid label {
  errs += "no dots"
else 
  errs += regex
}
```

This would make nicer validation errors, too.  Result something like:

```
Warning: metadata.name: this is used in Pod names and hostnames, which can result in surprising behavior: a DNS label is recommended: [must not contain dots]
```

or 

```
Warning: metadata.name: this is used in Pod names and hostnames, which can result in surprising behavior: a DNS label is recommended: [must be no more than 63 characters]
```

or

```
Warning: metadata.name: this is used in Pod names and hostnames, which can result in surprising behavior: a DNS label is recommended: [must be no more than 63 characters, must not contain dots]
```

Or we could just do this in the warning logic and get the shorter errors you wrote

liggitt replied:
I'm happy with any of those options... making the validation methods spit out nicer errors for everyone seems like an improvement, but one that will probably punish you by making you go change tests expecting specific nasty regex errors. Your call.

thockin replied:
```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go b/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go
index e767092dd87..0b8a6cb354a 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go
@@ -191,7 +191,13 @@ func IsDNS1123Label(value string) []string {
                errs = append(errs, MaxLenError(DNS1123LabelMaxLength))
        }
        if !dns1123LabelRegexp.MatchString(value) {
-               errs = append(errs, RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
+               if dns1123SubdomainRegexp.MatchString(value) {
+                       // It was a valid subdomain and not a valid label.  Since we
+                       // already checked length, it must be dots.
+                       errs = append(errs, "must not contain dots")
+               } else {
+                       errs = append(errs, RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
+               }
        }
        return errs
 }
```

thockin replied:
pushed, PTAL

liggitt commented:
This PR is marked as fixing https://github.com/kubernetes/kubernetes/issues/104195, but it doesn't look like it touches statefulset... does it really fix that issue or is it just related?

Are statefulsets more impacted by this than the other types this PR is warning about (to the point where it might deserve an actual validation error on statefulset)?

thockin commented:
Stefulset was fixed by https://github.com/kubernetes/kubernetes/pull/114172

sftim commented:
@thockin, I can't work out what you intend for the release note text to be here. This change will need a short release note summary and would benefit from some extra notes to accompany the summary.

We could also update https://kubernetes.io/docs/concepts/workloads/controllers/deployment/ and similar pages to include advice so that an informed reader doesn't see such a warning.
If this merges, please consider filing an issue against k/website, so that we can track the need to document this for the upcoming release.

thockin commented:
@sftim I thought the release-note I wrote was pretty clear - help me clarify?

sftim commented:
What's the _change_ here? I can guess from the PR title, but I'm not sure that the difference between before and after will be obvious to someone reading just the changelog.

It's OK to provide an example of the new behavior but we first need to explain the difference between before and afterwards.

thockin commented:
"""
Pods, ReplicationControllers, ReplicaSets, Deployments, Jobs, and CronJobs allow their `metadata.name` field to be a DNS Subdomain (e.g. foo.bar), which can produce unexpected behaviors when used as a pod's hostname.  **Creating an instance of these types with name that is not a DNS Label (e.g. foobar, foo-bar) will get API warnings recommending the stricter format.**  All existing API operations are still allowed - these are only warnings.  
"""

Better to lead with the behavior change?

"""
Creating Pods, ReplicationControllers, ReplicaSets, Deployments, Jobs, and CronJobs whose `metadata.name` field is not a DNS Label (e.g. foobar, foo-bar) will now get API warnings.  These types allow their `metadata.name` field to be a DNS Subdomain (e.g. foo.bar), but this can produce unexpected behaviors when used as a pod's hostname.  All existing API operations are still allowed - these are only warnings.
"""

sftim commented:
I'd consider:
```markdown
Creating Pods, ReplicaSets, Deployments, Jobs, CronJobs and ReplicationControllers whose `metadata.name`
field is *not* a valid DNS label will now get API warnings. For example, naming a CronJob `foo.bar.example`
now returns a warning. Naming a CronJob `foobar` or `foo-bar-example` does not return a warning because
those names are valid DNS labels.
The Kubernetes API for these API kinds requires that the `metadata.name` field is a DNS subdomain
name. All DNS label names are valid DNS subdomain names, but using a DNS subdomain that is not also
a valid label can produce unexpected behaviors when used as a pod's hostname.
All existing API operations are still allowed - these are only warnings.
```
_or_ write a summary and use the option for further details:
```markdown
Creating Pods, ReplicaSets, Deployments, Jobs, CronJobs and ReplicationControllers whose
`metadata.name` field is *not* a valid DNS label will now get API warnings.
All existing API operations are still allowed - these are only warnings.
```
```markdown
For example, naming a CronJob `foo.bar.example` now returns a
[warning](https://blog.k8s.io/2020/09/03/warnings/).
Naming a CronJob `foobar` or `foo-bar-example` does not return a warning because
those names are valid DNS labels. The Kubernetes API for these API kinds enforces that the
`metadata.name` field is a DNS _subdomain_ name. All DNS label names are valid DNS
subdomain names, but using a DNS subdomain that is not also a valid label can produce
unexpected behaviors when used as a pod's hostname.

See [Object Names and IDs](https://k8s.io/docs/concepts/overview/working-with-objects/names/)
for more information.
```

I don't know the exact mechanism for changelogs but I've seen that there is space for adding additional details to a changelog entry, and this second option feels like a good way to use that.

Finally, what about StatefulSets?

liggitt commented:
> Finally, what about StatefulSets?

statefulsets were handled differently and separately in https://github.com/kubernetes/kubernetes/pull/114172

thockin commented:
PTAL at relnote

thockin committed:
Change DNS Label validation to check for dots

This will produce a better error message for the more common case of
using a DNS subdomain where a label is needed.

liggitt commented:
/lgtm

sftim commented:
For these release notes, I'd expect to see two blocks. One block that starts with starting <tt>\`\`\`release-note</tt> block and another starting <tt>\`\`\`docs</tt>. That second block is there to give us room for more info.

Something like:
<pre>
```release-note
Added warnings about workload names that are not valid DNS labels.
```
```docs
Creating Pods, ReplicaSets, Deployments, Jobs, CronJobs, or ReplicationControllers whose
`metadata.name` field is *not* a valid
[DNS label](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names)
will now get [API warnings](https://blog.k8s.io/2020/09/03/warnings/).
All previously valid names and API operations are still allowed - these are only warnings.

For example, naming a Deployment `foo.bar` (a DNS subdomain) now returns a  warning,
whereas `foobar` or `foo-bar` (DNS labels) do not.  For backwards compatibility, the Kubernetes
API for these resources allows the `metadata.name` field to be a
[DNS subdomain](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
 DNS labels are a subset of DNS subdomains.  All DNS label names are valid DNS subdomain names,
but using a DNS subdomain that is not _also_ a valid DNS label (e.g. a name which includes dots)
can produce unexpected behaviors when used as a pod's hostname.

See [Object Names and IDs](https://k8s.io/docs/concepts/overview/working-with-objects/names/)
for more information.
```
</pre>

Splitting in a different place is OK. Our [guidance](https://www.kubernetes.dev/docs/guide/release-notes/#writing-good-release-notes) doesn't include an example with additional docs so I'm not sure what represents good practise.
I also don't know where the information in the `docs` part ends up.


Also, in that example, I changed “All existing names“ to “**All previously valid names**”; I wanted to make it more obvious that any **create** that might previously have been acceptable is still acceptable after this change. I don't want readers to wrongly infer that only objects that existed before the upgrade are exempt from the validation change.

thockin commented:
Edited as suggested, we'll see what release folks think as the release notes get assembled. :)

sftim commented:
Thanks for all the edits!

cici37 commented:
/triage accepted
